Userland/Applications/SystemMonitor/ThreadStackWidget.cpp

```diff
@@ -121,7 +121,7 @@ void ThreadStackWidget::refresh()
         [weak_this = make_weak_ptr()](auto result) {
             if (!weak_this)
                 return;
-            Core::EventLoop::main().post_event(const_cast<Core::Object&>(*weak_this), make<CompletionEvent>(move(result)));
+            Core::EventLoop::main().with_locked([&](auto* main_event_loop) { main_event_loop->post_event(const_cast<Core::Object&>(*weak_this), make<CompletionEvent>(move(result))); });
```

Comment:
Hmm, this feels very clunky having to used `with_locked()` outside of `Core::EventLoop` itself.
Can we come up with a public API that hides this detail?

Reply:
The reason this needs to exist in some form or another is that inside the lambda we're protected by the mutex lock. Passing the raw object itself can lead to unsynchronized accesses. Passing a MutexLocker out of the function is something where I don't really understand the behavior well. Would we temporarily lock and unlock, then re-lock the mutex lock? What if the caller stores the MutexLocker on the heap? I figured that this was the best solution, especially given that there's just one user of the API right now.

Reply:
Why can't `post_event()` itself be responsible for locking the mutex?

Reply:
Because at that point we're inside the EventLoop which doesn't know that it's synchronized and/or needs to be locked.

Reply:
Hmm, I don't understand the difference here.
You've already dereferenced a pointer to the event loop by accessing `Core::EventLoop::main()`. At that point, taking the lock outside or inside `Core::EventLoop::post_event()` seems equivalent.
If the issue is that you don't want to access `main()` unlocked, then what you really need is something like a static `Core::EventLoop::with_main_locked(callback)`, no?